src/benchmarklib/benchmark_runner.cpp

```diff
@@ -103,52 +111,157 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
   std::random_device random_device;
   std::mt19937 random_generator(random_device());
 
-  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
-  while (state.keep_running()) {
-    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
-
-    for (const auto& named_query : mutable_named_queries) {
-      const auto query_benchmark_begin = std::chrono::steady_clock::now();
-
-      // Execute the query, we don't care about the results
-      _execute_query(named_query);
-
-      const auto query_benchmark_end = std::chrono::steady_clock::now();
-
-      auto& query_benchmark_result = _query_results_by_query_name[named_query.first];
-      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
-      query_benchmark_result.num_iterations++;
+  const auto number_of_queries = mutable_named_queries.size();
+
+  // The atomic uints are modified by other threads when finishing a query set, to keep track of when we can
+  // let a simulated client schedule the next set, as well as the total number of finished query sets so far
+  auto currently_running_clients = std::atomic_uint{0};
+  auto finished_query_set_runs = std::atomic_uint{0};
+  auto finished_queries_total = std::atomic_uint{0};
+
+  auto tasks = std::vector<std::shared_ptr<AbstractTask>>{};
+  auto state = BenchmarkState{_config.max_duration};
+
+  while (state.keep_running() && finished_query_set_runs.load(std::memory_order_relaxed) < _config.max_num_query_runs) {
+    // We want to only schedule as many query sets simultaneously as we have simulated clients
+    if (currently_running_clients.load(std::memory_order_relaxed) < _config.clients) {
+      currently_running_clients++;
+      std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+
+      for (const auto& named_query : mutable_named_queries) {
+        // The on_query_done callback will be appended to the last Task of the query,
+        // to measure its duration as well as signal that the query was finished
+        const auto query_run_begin = std::chrono::steady_clock::now();
+        auto on_query_done = [query_run_begin, named_query, number_of_queries, &currently_running_clients,
+                              &finished_query_set_runs, &finished_queries_total, &state, this]() {
+
+          if (finished_queries_total++ % number_of_queries == 0) {
+            currently_running_clients--;
+            finished_query_set_runs++;
+          }
+
+          if (!state.is_done()) { // To prevent queries to add their results after the time is up
```

Comment:
I added this, because after the time is up we still "wait" for the currently running clients to finish, but of course we don't want them to add their results anymore.